crates/common/src/provider/runtime_transport.rs

```diff
@@ -211,19 +211,31 @@ impl RuntimeTransport {
     pub fn request(&self, req: RequestPacket) -> TransportFut<'static> {
         let this = self.clone();
         Box::pin(async move {
-            if this.inner.read().await.is_none() {
-                let mut inner = this.inner.write().await;
-                *inner = Some(this.connect().await.map_err(TransportErrorKind::custom)?)
+            let mut inner = this.inner.read().await;
+            if inner.is_none() {
+                drop(inner);
+                let mut inner_mut = this.inner.write().await;
+                if inner_mut.is_none() {
+                    *inner_mut = Some(this.connect().await.map_err(TransportErrorKind::custom)?);
+                }
+                drop(inner_mut);
+                inner = this.inner.read().await;
             }
 
-            let mut inner = this.inner.write().await;
             // SAFETY: We just checked that the inner transport exists.
-            let inner_mut = inner.as_mut().expect("We should have an inner transport.");
-
-            match inner_mut {
-                InnerTransport::Http(http) => http.call(req),
-                InnerTransport::Ws(ws) => ws.call(req),
-                InnerTransport::Ipc(ipc) => ipc.call(req),
+            match inner.as_ref().unwrap() {
```

Comment:
Another pedantic style nit: I prefer having `expect()` with a message instead of a silent unwrap, just to make it super clear when an error happens